masq_lib/src/percentage.rs

```diff
@@ -1,77 +1,226 @@
+        N: PercentageInteger,
+        <N as TryFrom<i8>>::Error: Debug,
+        i16: TryFrom<N>,
+        <i16 as TryFrom<N>>::Error: Debug,
     {
-        let zero = N::from(0);
-        if num == zero || N::from(self.per_cent) == zero {
+        if let Some(zero) = self.return_zero(num) {
             return zero;
         }
 
-        let a = match N::from(self.per_cent).checked_mul(&num) {
+        let product_before_final_div = match N::try_from(self.degree as i8)
+            .expect("Each type has 100")
+            .checked_mul(&num)
+        {
             Some(num) => num,
             None => return self.handle_upper_overflow(num),
         };
 
-        if a < N::from(10) {
-            return N::from(0);
-        }
+        Self::div_by_100_and_round(product_before_final_div)
+    }
 
-        let rounding = if Percentage::should_be_rounded_down(a) {
-            N::from(0)
+    fn return_zero<N>(&self, num: N) -> Option<N>
+    where
+        N: PercentageInteger,
+        <N as TryFrom<i8>>::Error: Debug,
+    {
+        let zero = N::try_from(0).expect("Each type has 0");
+        if num == zero || N::try_from(self.degree as i8).expect("Each type has 100") == zero {
+            Some(zero)
         } else {
-            N::from(1)
-        };
+            None
+        }
+    }
+
+    fn div_by_100_and_round<N>(num: N) -> N
+    where
+        N: PercentageInteger,
+        <N as TryFrom<i8>>::Error: Debug,
+    {
+        let divisor = N::try_from(100).expect("Each type has 100");
+        let desired_rounding = Self::should_be_rounded_to(num, divisor);
+        let significant_digits_only = num.checked_div(&divisor).expect("Division failed");
+
+        macro_rules! adjust_num {
+            ($significant_digits: expr, $method_add_or_sub: ident, $msg_in_expect: literal) => {
+                $significant_digits
+                    .$method_add_or_sub(&N::try_from(1).expect("Each type has 1"))
+                    .expect($msg_in_expect)
+            };
+        }
+
+        match desired_rounding {
+            RoundingTo::BiggerPositive => {
+                adjust_num!(significant_digits_only, checked_add, "Addition failed")
+            }
+            RoundingTo::BiggerNegative => {
+                adjust_num!(significant_digits_only, checked_sub, "Subtraction failed")
+            }
+            RoundingTo::SmallerNegative | RoundingTo::SmallerPositive => significant_digits_only,
+        }
+    }
 
-        let hundred = N::from(100);
+    fn should_be_rounded_to<N>(num: N, divisor: N) -> RoundingTo
+    where
+        N: PercentageInteger,
+        <N as TryFrom<i8>>::Error: Debug,
+    {
+        let least_significant_digits: N = num % divisor;
+        let is_signed = num < N::try_from(0).expect("Each type has 0");
+        let divider = N::try_from(50).expect("Each type has 50");
+        let abs_of_significant_digits =
+            Self::abs_of_least_significant_digits(least_significant_digits, is_signed);
```

Comment:
You'd likt o fix the name of this variable to contain the word `least`, as it's present in the fn name but missing in the variable name. Also, if you could rename `is_signed` to `is_negative` it'll be more explicit. I didn't mean to say, `is_signed` is wrong. So, please change it accordingly.